### PR TITLE
[FSDP] Improve support for CPU tensors.

### PR DESCRIPTION
**Pull request by kumpera**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #103214
* #103172
* __->__ #103171

Don't emit device index when using CPU devices.
Don't call Tensor::record_stream as it's CUDA only op.

### CODE REVIEW

File: torch/distributed/fsdp/_shard_utils.py [hunk 1]
```diff
@@ -102,7 +109,7 @@ def _create_chunk_sharded_tensor(
     chunk_offsets = [[d0] + offsets for d0 in dim0_offsets]
     device_type = distributed_c10d._get_pg_default_device(pg).type
     placements = [
-        f"rank:{r}/{device_type}:{r % num_devices_per_node}"
+        _get_remove_device_str(r, device_type, num_devices_per_node)
```
Comment (rohan-varma) [inline torch/distributed/fsdp/_shard_utils.py:112]: get remote?

**Review by rohan-varma — APPROVED** [review]

**Review by wz337 — APPROVED** [review]
LGTM!

File: torch/distributed/fsdp/_shard_utils.py (hunk 1 above)
Comment (kumpera) [reply torch/distributed/fsdp/_shard_utils.py:112]: I suck at naming. I'l remove the `_get`

Comment (kumpera) [reply torch/distributed/fsdp/_shard_utils.py:112]: The issue is that we can't have "cpu:1"

**Pull request closed without merging**